src/net_processing.cpp

```diff
@@ -2777,11 +2798,34 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         std::vector<CAddress> vAddrOk;
         int64_t nNow = GetAdjustedTime();
         int64_t nSince = nNow - 10 * 60;
+
+        // Update/increment addr rate limiting bucket.
+        const auto current_time = GetTime<std::chrono::microseconds>();
+        if (peer->m_addr_token_bucket < MAX_ADDR_PROCESSING_TOKEN_BUCKET) {
+            // Don't increment bucket if it's already full
+            const auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
+            const double increment = CountSecondsDouble(time_diff) * MAX_ADDR_RATE_PER_SECOND;
+            peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_PROCESSING_TOKEN_BUCKET);
```

Comment:
Here we can calculate the number of to be dropped addresses and also log the message, for example:

```diff
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -2796,36 +2796,46 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
 
         // Store the new addresses
         std::vector<CAddress> vAddrOk;
         int64_t nNow = GetAdjustedTime();
         int64_t nSince = nNow - 10 * 60;
 
+        const bool rate_limited = !pfrom.HasPermission(NetPermissionFlags::Addr);
+        uint64_t num_proc = 0, num_rate_limit = 0;
+
         // Update/increment addr rate limiting bucket.
         const auto current_time = GetTime<std::chrono::microseconds>();
         if (peer->m_addr_token_bucket < MAX_ADDR_PROCESSING_TOKEN_BUCKET) {
             // Don't increment bucket if it's already full
             const auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
             const double increment = CountSecondsDouble(time_diff) * MAX_ADDR_RATE_PER_SECOND;
             peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_PROCESSING_TOKEN_BUCKET);
+
+            if (rate_limited && vAddr.size() > peer->m_addr_token_bucket) {
+                num_rate_limit =
+                    static_cast<uint64_t>(std::ceil(vAddr.size() - peer->m_addr_token_bucket));
+                LogPrint(BCLog::NET,
+                         "Rate limiting %u of %u addresses from a message from peer=%d%s\n",
+                         pfrom.GetId(),
+                         num_rate_limit,
+                         vAddr.size(),
+                         fLogIPs ? ", peeraddr=" + pfrom.addr.ToString() : "");
+            }
         }
         peer->m_addr_token_timestamp = current_time;
 
         Shuffle(vAddr.begin(), vAddr.end(), FastRandomContext());
-        const bool rate_limited = !pfrom.HasPermission(NetPermissionFlags::Addr);
-        uint64_t num_proc = 0, num_rate_limit = 0;
         for (CAddress& addr : vAddr)
         {
             if (interruptMsgProc)
                 return;
 
             // Apply rate limiting.
             if (rate_limited) {
                 if (peer->m_addr_token_bucket < 1.0) {
-                    LogPrint(BCLog::NET, "Rate limiting an address message from peer=%d\n", pfrom.GetId());
-                    num_rate_limit++;
-                    continue;
+                    break;
                 }
                 peer->m_addr_token_bucket -= 1.0;
             }
             // We only bother storing full nodes, though this may include
             // things which we would not make an outbound connection to, in
             // part because we may make feeler connections to them.
```

Reply:
Now I see that I have messed up the order of the arguments to `LogPrint()` :facepalm:. Correction:

```diff
                 LogPrint(BCLog::NET, "Rate limiting %u of %u addresses from a message from peer=%d%s\n
-                         pfrom.GetId(),
                          num_rate_limit,
                          vAddr.size(),
+                         pfrom.GetId(),
                          fLogIPs ? ", peeraddr=" + pfrom.addr.ToString() : "");
```

Reply:
This better be outside of the `if`:

```diff
         if (peer->m_addr_token_bucket < MAX_ADDR_PROCESSING_TOKEN_BUCKET) {
             // Don't increment bucket if it's already full
             const auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
             const double increment = CountSecondsDouble(time_diff) * MAX_ADDR_RATE_PER_SECOND;
             peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_AD
-            if (rate_limited && vAddr.size() > peer->m_addr_token_bucket) {
-                num_rate_limit = (uint64_t)std::ceil(vAddr.size() - peer->m_addr_token_bucket);
-                LogPrint(BCLog::NET, "Rate limiting %u of %u addresses from a message from peer=%d%s\n
-                         pfrom.GetId(),
-                         num_rate_limit,
-                         vAddr.size(),
-                         fLogIPs ? ", peeraddr=" + pfrom.addr.ToString() : "");
-            }
         }
+        if (rate_limited && vAddr.size() > peer->m_addr_token_bucket) {
+            num_rate_limit = (uint64_t)std::ceil(vAddr.size() - peer->m_addr_token_bucket);
+            LogPrint(BCLog::NET,
+                     "Rate limiting %u of %u addresses from a message from peer=%d%s\n",
+                     num_rate_limit,
+                     vAddr.size(),
+                     pfrom.GetId(),
+                     fLogIPs ? ", peeraddr=" + pfrom.addr.ToString() : "");
+        }
         peer->m_addr_token_timestamp = current_time;
```

Reply:
Done, and moved using @jnewbery's suggestion below.